bazel/linux/test.bzl

```diff
@@ -4,6 +4,83 @@ load("//bazel/utils:macro.bzl", "mconfig", "mcreate_rule")
 load("//bazel/utils:exec_test.bzl", "exec_test")
 load("//bazel/linux:bundles.bzl", "kunit_bundle")
 
+def _test_runner(ctx):
+    script_begin = """#!/bin/bash
+set -eu -o pipefail
+GREEN="\033[32m"
+RED="\033[31m"
+RESET="\033[0m"
+failed_outputs=()
+tests=0
+failed=0
+run_test() {
+    local title="$1"
+    local script="$2"
+    local success=0
+    local output
+    local status
+    local start_time_ns
+    local stop_time_ns
+    local duration_ns
+
+    printf "%-73s" "${title}"
+    start_time_ns=$(($(date +%s%N)))
+
+    output=$("${script}" 2>&1) && success=1
+
+    tests=$[tests + 1]
+    if [ $success -eq 1 ]; then
+        status="${GREEN}PASSED${RESET}"
+    else
+        failed=$[failed + 1]
+        status="${RED}FAILED${RESET}"
+        failed_outputs+="==================== Test output for ${title}:\\n${output}\\n"
+    fi
+    stop_time_ns=$(($(date +%s%N)))
+    duration_ns=$(($stop_time_ns - $start_time_ns))
+    printf "%s in %.1fs\\n" "$status" "$duration_ns""e-9"
+}
+"""
+
+    script_end = """
+for o in "${failed_outputs[@]}"; do
+    echo -ne "$o"
+done
+if [ $failed -ne 0 ]; then
+    echo "================================================================================"
+fi
+succeeded=$[tests - failed]
+echo "Executed ${tests} test(s): ${succeeded} test(s) pass(es) and ${failed} fail(s)."
+exit $failed
+"""
+
+    tests = []
+    runfiles = ctx.runfiles()
+    for test in ctx.attr.tests:
+        binary = test.files_to_run.executable
+        tests.append("run_test {title} {script}".format(
+            title = test.label,
+            script = binary.short_path,
+        ))
+        runfiles = runfiles.merge(ctx.runfiles([binary]))
+
+    script = ctx.actions.declare_file("{}_test_runner.sh".format(ctx.attr.name))
+    ctx.actions.write(script, script_begin + "\n".join(tests) + script_end)
+    print(script_begin + "\n".join(tests) + script_end)
+    return [DefaultInfo(runfiles = runfiles, executable = script)]
+
+test_runner = rule(
+    doc = """TODO""",
+    implementation = _test_runner,
+    executable = True,
+    attrs = {
+        "tests": attr.label_list(
+            mandatory = True,
+            doc = "List of executable targets to run in the emulator as tests. The duration and the exit status of each test is recorded and reported.",
+        ),
+    },
+)
+
 def qemu_test(name, kernel_image, run, qemu_binary = None,
```

Comment:
I think we said we'd add a separate setup step here? separate PR?

Reply:
Added.